File: algorithms/TwoSum.js
function twoSum(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    let pairIndex = nums.indexOf(target - nums[i], i + 1);
    if (pairIndex !== -1)
      return [i, pairIndex];
  }
  return null;
};

console.log(
  twoSum([2,7,11,15], 9), // [0,1]
  twoSum([3,2,4], 6),     // [1,2]
  twoSum([3,3], 6),       // [0,1]
  twoSum([3,3], 9),       // null
)

File: algorithms/TwoSum.go
package algorithms

// return indices
func TwoSum(nums []int, target int) []int {
	numsMap := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		numsMap[nums[i]] = i
	}

	for i := 0; i < len(nums); i++ {
		pair := target - nums[i]
		pairIndex, pairExists := numsMap[pair]
		if pairExists && pairIndex != i {
			return []int{i, pairIndex}
		}
	}
	return nil
}

// return values
func TwoSumValues(nums []int, target int) []int {
	res := TwoSum(nums, target)
	if res != nil {
		return []int{nums[res[0]], nums[res[1]]}
	}
	return nil
}
